devonfw/ide-urls#17: fix UrlUpdater to auto-remove broken URLs and allow self-repair

hohwille opened the pull request:
fixes devonfw/ide-urls#17

or at least tries to fix the problem. Releases that are gone cannot come back but they will now be deleted if 404 have been detected for 14+ days.
If no valid URL remains for a version the entire version will be gone.
Therefore it will be recreated if it still exists (e.g. on different mirror).

I found and fixed many evil bugs and spent a long time debugging this mess...

hohwille commented on cli/src/main/java/com/devonfw/tools/ide/url/model/file/UrlStatusFile.java:
```diff
@@ -70,11 +70,23 @@ protected void doLoad() {
   protected void doSave() {
 
-    try (BufferedWriter writer = Files.newBufferedWriter(getPath(), StandardOpenOption.CREATE)) {
+    try (BufferedWriter writer = Files.newBufferedWriter(getPath())) {
```
This was an evil bug. If explicit `StandardOpenOperation`s are specified as argument, I usually think that the developer made some smart choice here. However, this was pure non-sense:
If the JSON file already existed, the new JSON was appended causing invalid JSON and reading the JSON diff in git made it very tricky to figure out what is actually going wrong here.

hohwille commented on cli/src/main/java/com/devonfw/tools/ide/url/model/file/json/StatusJson.java:
```diff
@@ -64,7 +66,37 @@ public void setUrls(Map<Integer, UrlStatus> urlStatuses) {
+    Iterator<Entry<Integer, UrlStatus>> entryIterator = this.urls.entrySet().iterator();
+    while (entryIterator.hasNext()) {
+      Entry<Integer, UrlStatus> entry = entryIterator.next();
+      if (!entry.getValue().checkStillUsed()) {
```
unused entries were never removed before this change ending up as dead JSON causing confusion.
Really hard to understand what was going on since the JSON keys are only cryptic hashes of the URLs.
So if you have 5 URLs and 11 entries in JSON you get puzzled but have no idea which entry was right and which is wrong. Some entries were (are currently) entirely empty so they neither contain `success` nor `error`.

hohwille commented on cli/src/main/java/com/devonfw/tools/ide/url/model/file/json/UrlStatus.java:
```diff
@@ -48,4 +50,29 @@ public void setError(UrlStatusState error) {
+    return (this.error == null) && (this.success == null);
+  }
+
+  public boolean checkStillUsed() {
```
If you use a regular getter (`isStillUsed()`), then Jackson will write this property into JSON.

hohwille commented on cli/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:
```diff
@@ -662,36 +746,46 @@ protected void updateExistingVersions(UrlEdition edition) {
+        Set<String> urls = new HashSet<>(urlDownloadFile.getUrls());
         for (String url : urls) {
-          if (shouldVerifyDownloadUrl(version, statusJson, toolWithEdition, now)) {
+          if (shouldVerifyDownloadUrl(url, statusJson, toolWithEdition, now)) {
```
very evil bug: first argument is `url` but `version` was given as 1st arg before this change. 👿

hohwille commented on cli/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:
```diff
@@ -303,28 +315,23 @@ protected boolean isSuccess(HttpResponse<?> response) {
+          toolWithEdition, version, url, checksum, existingChecksum);
       return false;
     } else {
-      urlDownloadFile.addUrl(url);
```
bad design: the method name `isChecksumStillValid` implies that we only check the checksum here and not do such side-effects.

hohwille commented on cli/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:
```diff
@@ -377,43 +384,50 @@ protected boolean isValidContentType(String contentType) {
-      if (urlChecksum != null) {
-        logger.warn("Checksum is already existing for: {}, skipping.", url);
-        doUpdateStatusJson(success, statusCode, edition, urlVersion, url, true);
-        return true;
```
evil bug: here the `success` status was lost and always `true` was returned even if the head requested resulted in an invalid response.

hohwille commented on cli/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:
```diff
@@ -377,43 +384,50 @@ protected boolean isValidContentType(String contentType) {
-      success = isChecksumStillValid(url, urlVersion, os, architecture, checksum, tool, version);
-    }
-
-    if (success) {
```
confusing code: two `if (success) { ... }` statements in a row (without `else`). If the first did not match, the second cannot match either.

hohwille replied:
However, I first caused another bug with this change so I had to update again...

hohwille commented on cli/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:
```diff
@@ -511,7 +528,7 @@ private void doUpdateStatusJson(boolean success, int statusCode, String edition,
 
     if (success) {
-      boolean setSuccess = !update;
+      boolean setSuccess = !update || (successStatus == null);
```
bug: if `successStatus` is `null`, we always need to update success timestamp.

hohwille commented on cli/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:
```diff
@@ -551,11 +586,54 @@ private void doUpdateStatusJson(boolean success, int statusCode, String edition,
+  }
+
+  private static boolean isErrorCodeForAutomaticUrlRemoval(Integer code) {
+    return Integer.valueOf(404).equals(code);
```
here we could include even more HTTP status codes like `500` or `410`.
However, this one is the really relevant one...

alfeilex replied:
In the raw logs, the only other status code for an error was `503`. So `404` should be enough.

hohwille replied:
Thanks for the feedback.
We can still extend this here after the merge if we face new situations.
Hence resolving this.

hohwille commented on cli/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:
```diff
@@ -551,11 +586,54 @@ private void doUpdateStatusJson(boolean success, int statusCode, String edition,
+    return Integer.valueOf(404).equals(code);
+  }
+
+  private boolean isGreaterThan(Duration errorDuration, Duration daysUntilDeletionOfBrokenUrl) {
```
move to `DateTimeUtil`? Is this relevant for reuse?

hohwille replied:
We can still do that later if we think it is of value.

hohwille commented on cli/src/main/resources/logback.xml:
```diff
@@ -8,8 +8,8 @@
   </appender>
 
-  <root level="INFO">
+  <root>
```
Intellij complaint about this and said this attribute was not defined. However, this change caused massive log-spam.
```suggestion
  <!-- Setting the default threshold to INFO to avoid logspam. Please ignore IntelliJ lies that the attribute level is undefined. -->
  <root level="INFO">
```

hohwille commented:
Tests still failing:

> [ERROR] Failures: 
[ERROR]   UrlUpdaterReportTest.testReportAfterVersionForOsRemoved:104 
expected: mocked/mocked:addVersionSuccess=1, addVersionFailure=0, verificationSuccess=7, verificationFailure=8
 but was: mocked/mocked:addVersionSuccess=1, addVersionFailure=0, verificationSuccess=2, verificationFailure=2
[ERROR]   UrlUpdaterReportTest.testReportAfterVersionsForRemovedAndReversedUrlResponses:143 
expected: mocked/mocked:addVersionSuccess=1, addVersionFailure=0, verificationSuccess=6, verificationFailure=6
 but was: mocked/mocked:addVersionSuccess=1, addVersionFailure=0, verificationSuccess=2, verificationFailure=2
[ERROR]   UrlUpdaterReportTest.testReportWithExistVersionsAndFailedDownloads:75 
expected: mocked/mocked:addVersionSuccess=0, addVersionFailure=0, verificationSuccess=6, verificationFailure=6
 but was: mocked/mocked:addVersionSuccess=0, addVersionFailure=0, verificationSuccess=0, verificationFailure=0
[INFO] 
[ERROR] Tests run: 355, Failures: 3, Errors: 0, Skipped: 0

jan-vcapgemini commented on cli/src/main/java/com/devonfw/tools/ide/url/model/folder/AbstractUrlFolder.java:
```diff
@@ -61,6 +67,45 @@ public C getOrCreateChild(String name) {
+    }
+  }
+
+  private static void delete(Path path) {
+
+    LOG.debug("Deleting {}", path);
+    if (Files.exists(path)) {
+      try {
+        deleteRecursive(path);
+      } catch (IOException e) {
+        throw new RuntimeException("Failed to delete " + path);
+      }
+    } else {
+      LOG.warn("Could not delete file {} because it does not exist.", path);
+    }
+  }
```
Move to FileAccess?

hohwille replied:
Cannot until we have solved #404 since currently `FileAccessImpl` is based on `IdeContext` and uses that for logging.

jan-vcapgemini commented on cli/src/main/java/com/devonfw/tools/ide/url/model/folder/AbstractUrlFolder.java:
```diff
@@ -61,6 +67,45 @@ public C getOrCreateChild(String name) {
+    }
+  }
+
+  private static void deleteRecursive(Path path) throws IOException {
+
+    if (Files.isDirectory(path)) {
+      try (Stream<Path> childStream = Files.list(path)) {
+        Iterator<Path> iterator = childStream.iterator();
+        while (iterator.hasNext()) {
+          Path child = iterator.next();
+          deleteRecursive(child);
+        }
+      }
+    }
+    LOG.trace("Deleting {}", path);
+    Files.delete(path);
+
+  }
```
Move to FileAccess?

hohwille replied:
same here.

jan-vcapgemini commented on cli/src/main/java/com/devonfw/tools/ide/url/model/folder/AbstractUrlFolder.java:
```diff
@@ -87,6 +132,18 @@ public Set<String> getChildNames() {
+  /**
+   * Will discard all children and reload from the disc.<br>
+   * <br>
+   * <b>ATTENTION:</b> This method is only for a very specific use-case and should only be called when you know exactly what you are doing.
```
Please add use case example for this method.

hohwille replied:
removed. Was used before but I then came to another solution. Could be helpful later but I can then readd it whenever needed.

jan-vcapgemini commented on cli/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:
```diff
@@ -653,7 +738,12 @@ protected void updateExistingVersions(UrlEdition edition) {
+          if (urlVersion.getChildren().isEmpty()) {
+            logger.warn("Finally deleting broken or disappeared version {}", urlVersion.getPath());
+            urlVersion.delete();
+          } else {
+            urlVersion.save(); // why do we call this here?
+          }
```
Add TODO for advanced debugging if this is needed.

jan-vcapgemini commented on cli/src/main/java/com/devonfw/tools/ide/url/updater/UpdateManager.java:
```diff
@@ -94,14 +94,31 @@ public void updateAll() {
+    }
+  }
+
+  public void update(String tool) {
```
Please add Javadoc or Override

hohwille replied:
Done. Only for local dev actually and quite self-explanatory but JavaDoc is always better to have for public stuff.

jan-vcapgemini reviewed (changes requested):
Thank you for your bugfixes. We've added some change requests and questions.
Co-Reviewed by: @alfeilex @fahamzaoui

hohwille commented:
> Thank you for your bugfixes. We've added some change requests and questions.

I reworked that already and IMHO all review feedback is addressed.
Did you test this already?
Do you have any hint how to proceed with the failing test? Is there a bug or is the test incorrect?

I would like to see what happens after we merged this and the UrlUpdater action ran again... Will this fix our problems or not.

hohwille commented on cli/src/main/java/com/devonfw/tools/ide/url/model/file/json/UrlStatus.java:
```diff
@@ -48,4 +50,32 @@ public void setError(UrlStatusState error) {
+  /**
+   * Sets {@link #checkStillUsed()} to {@code true}.
+   */
+  void markSillUsed() {
```
Done.

fahamzaoui commented on cli/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:
```diff
@@ -377,43 +384,51 @@ protected boolean isValidContentType(String contentType) {
       SystemArchitecture architecture, String checksum) {
 
     HttpResponse<?> response = doCheckDownloadViaHeadRequest(url);
     int statusCode = response.statusCode();
```
please add nullpointer check here:
```
16:50:16.824 [main] - ERROR - c.d.t.ide.url.updater.JsonUrlUpdater - For tool gcviewer we failed to add version 1.35.
java.lang.NullPointerException: Cannot invoke "java.net.http.HttpResponse.statusCode()" because "response" is null
	at com.devonfw.tools.ide.url.updater.AbstractUrlUpdater.checkDownloadUrl(AbstractUrlUpdater.java:394)
```

hohwille replied:
Nice catch. But you found a bug that exists for long and was not introduced here.
Also interesting code/API:
```
  protected HttpResponse<?> doCheckDownloadViaHeadRequest(String url) {

    try {
      HttpRequest request = HttpRequest.newBuilder().uri(URI.create(url))
          .method("HEAD", HttpRequest.BodyPublishers.noBody()).timeout(Duration.ofSeconds(5)).build();

      return this.client.send(request, HttpResponse.BodyHandlers.ofString());
    } catch (Exception e) {
      logger.error("Failed to perform HEAD request of URL {}", url, e);
      return null;
    }
  }
```
Have you seen the stracktrace with the HTTP exception before?
If we want to handle `null` here, what shall we do with `statusCode`? IMHO I will set to `-1` in this case.

hohwille replied:
The `HttpResponse` is used all over the place. Adding `null` checks everywhere seems quite a challenge.
It would be nice to return a "mocked" `HttpResponse` instead of `null` but JDK API seems to lack the feature to create an instance of `HttpResponse` manually.
But if we assume the interface is stable, I could create a custom implementation and return that.

hohwille replied:
Kind of inconsistent:
```java
  protected HttpResponse<InputStream> doGetResponseAsStream(String url) {

    try {
      HttpRequest request = HttpRequest.newBuilder().uri(URI.create(url)).GET().build();
      return this.client.send(request, HttpResponse.BodyHandlers.ofInputStream());
    } catch (Exception e) {
      throw new IllegalStateException("Failed to retrieve response from url: " + url, e);
    }
  }
```

hohwille replied:
So I do not handle `null` now but prevent that `null` is returned:
https://github.com/devonfw/IDEasy/blob/8e724b6269c9ca5496797499203885bb5201f9f9/cli/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java#L492

fahamzaoui reviewed (commented):
found an issue with npe please fix.

fahamzaoui commented on cli/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:
```diff
@@ -662,22 +752,26 @@ protected void updateExistingVersions(UrlEdition edition) {
             HttpResponse<?> response = doCheckDownloadViaHeadRequest(url);
-            doUpdateStatusJson(isSuccess(response), response.statusCode(), edition, urlVersion, url, true);
-            modified = true;
+            doUpdateStatusJson(isSuccess(response), response.statusCode(), edition, urlVersion, url, urlDownloadFile, true);
```
another npe here broke whole Update process:
```
8.596 [main] - ERROR - c.d.t.i.u.updater.AbstractUrlUpdater - Failed to perform HEAD request of URL https://sourceforge.net/projects/gcviewer/files/gcviewer-1.35.jar
java.net.http.HttpConnectTimeoutException: HTTP connect timed out
	at java.net.http/jdk.internal.net.http.HttpClientImpl.send(HttpClientImpl.java:945)
	at java.net.http/jdk.internal.net.http.HttpClientFacade.send(HttpClientFacade.java:133)
	at com.devonfw.tools.ide.url.updater.AbstractUrlUpdater.doCheckDownloadViaHeadRequest(AbstractUrlUpdater.java:485)
	at com.devonfw.tools.ide.url.updater.AbstractUrlUpdater.updateExistingVersion(AbstractUrlUpdater.java:765)
	at com.devonfw.tools.ide.url.updater.AbstractUrlUpdater.updateExistingVersions(AbstractUrlUpdater.java:740)
	at com.devonfw.tools.ide.url.updater.JsonUrlUpdater.update(JsonUrlUpdater.java:47)
	at com.devonfw.tools.ide.url.UrlUpdateManagerTest.testUpdateGCView(UrlUpdateManagerTest.java:23)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:217)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:213)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:138)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:57)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:232)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)
Caused by: java.net.http.HttpConnectTimeoutException: HTTP connect timed out
	at java.net.http/jdk.internal.net.http.MultiExchange.toTimeoutException(MultiExchange.java:585)
	at java.net.http/jdk.internal.net.http.MultiExchange.getExceptionalCF(MultiExchange.java:532)
	at java.net.http/jdk.internal.net.http.MultiExchange.lambda$responseAsyncImpl$7(MultiExchange.java:452)
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934)
	at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:911)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
	at java.net.http/jdk.internal.net.http.common.SSLFlowDelegate.stopOnError(SSLFlowDelegate.java:1013)
	at java.base/java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:990)
	at java.base/java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:974)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
	at java.net.http/jdk.internal.net.http.common.SubscriberWrapper$DownstreamPusher.run1(SubscriberWrapper.java:295)
	at java.net.http/jdk.internal.net.http.common.SubscriberWrapper$DownstreamPusher.run(SubscriberWrapper.java:259)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$LockingRestartableTask.run(SequentialScheduler.java:182)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(SequentialScheduler.java:149)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:207)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:280)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:233)
	at java.net.http/jdk.internal.net.http.common.SubscriberWrapper.errorCommon(SubscriberWrapper.java:421)
	at java.net.http/jdk.internal.net.http.common.SSLFlowDelegate$Reader.errorCommon(SSLFlowDelegate.java:384)
	at java.net.http/jdk.internal.net.http.common.SubscriberWrapper.onError(SubscriberWrapper.java:412)
	at java.net.http/jdk.internal.net.http.SocketTube$InternalReadPublisher$ReadSubscription.signalCompletion(SocketTube.java:645)
	at java.net.http/jdk.internal.net.http.SocketTube$InternalReadPublisher$InternalReadSubscription.read(SocketTube.java:829)
	at java.net.http/jdk.internal.net.http.SocketTube$SocketFlowTask.run(SocketTube.java:181)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:207)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:280)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:233)
	at java.net.http/jdk.internal.net.http.SocketTube$InternalReadPublisher$InternalReadSubscription.signalError(SocketTube.java:778)
	at java.net.http/jdk.internal.net.http.SocketTube.signalClosed(SocketTube.java:159)
	at java.net.http/jdk.internal.net.http.PlainHttpConnection.close(PlainHttpConnection.java:428)
	at java.net.http/jdk.internal.net.http.AsyncSSLConnection.close(AsyncSSLConnection.java:116)
	at java.net.http/jdk.internal.net.http.Exchange$ConnectionAborter.closeConnection(Exchange.java:197)
	at java.net.http/jdk.internal.net.http.Exchange$ConnectionAborter.closeConnection(Exchange.java:180)
	at java.net.http/jdk.internal.net.http.Exchange.cancel(Exchange.java:288)
	at java.net.http/jdk.internal.net.http.MultiExchange.cancel(MultiExchange.java:265)
	at java.net.http/jdk.internal.net.http.ResponseTimerEvent.handle(ResponseTimerEvent.java:71)
	at java.net.http/jdk.internal.net.http.HttpClientImpl.purgeTimeoutsAndReturnNextDeadline(HttpClientImpl.java:1778)
	at java.net.http/jdk.internal.net.http.HttpClientImpl$SelectorManager.run(HttpClientImpl.java:1376)
Caused by: java.net.ConnectException: HTTP connect timed out
	at java.net.http/jdk.internal.net.http.MultiExchange.toTimeoutException(MultiExchange.java:586)
	... 38 common frames omitted

java.lang.IllegalStateException: Error while getting versions from JSON API https://api.github.com/repos/chewiebug/GCViewer/git/refs/tags

	at com.devonfw.tools.ide.url.updater.JsonUrlUpdater.update(JsonUrlUpdater.java:53)
	at com.devonfw.tools.ide.url.UrlUpdateManagerTest.testUpdateGCView(UrlUpdateManagerTest.java:23)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:217)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:213)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:138)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:57)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:232)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)
Caused by: java.lang.NullPointerException: Cannot invoke "java.net.http.HttpResponse.statusCode()" because "response" is null
	at com.devonfw.tools.ide.url.updater.AbstractUrlUpdater.updateExistingVersion(AbstractUrlUpdater.java:766)
	at com.devonfw.tools.ide.url.updater.AbstractUrlUpdater.updateExistingVersions(AbstractUrlUpdater.java:740)
	at com.devonfw.tools.ide.url.updater.JsonUrlUpdater.update(JsonUrlUpdater.java:47)
	... 69 more

Disconnected from the target VM, address: '127.0.0.1:55984', transport: 'socket'
```

hohwille replied:
Again thanks for finding this, but this is nothing broken by this PR but code that was broken from the start.
I think I need to create a dummy response instead of `null` in case of such HTTP exception.
